add-rtl8195am-feature-emac

M-ichae-l opened the pull request:
### Description
rtl8195am feature emac implementation.
coding style fix.
Updated emac greentea tests #6851.

### Pull request type
    [ ] Fix
    [ ] Refactor
    [ ] New target
    [X] Feature
    [ ] Breaking change

M-ichae-l committed:
add-rtl8195am-feature-emac

rtl8195am feature emac implementation.
coding style fix.
Updated emac greentea tests #6851.

kjbracey commented:
Looks like you've got a local copy of #6851 included - please rebase and tidy up.

M-ichae-l commented:
@kjbracey-arm 
ok. noted.

kjbracey commented on tools/test_configs/RealtekInterface.json:
```diff
@@ -9,7 +9,7 @@
         "connect-statement" : {
             "help" : "Disabled until EMAC updated",
-            "value" : null
+            "value" : "((RTWInterface *)net)->connect(WIFI_SSID, WIFI_PASSWORD, WIFI_SECURITY, WIFI_CHANNEL)"
```
Hadn't you needed an alternate version of this (`net->wiFiInterface()->connect`)?

M-ichae-l replied:
Yes. It should be `net->wiFiInterface()->connect`

kjbracey reviewed (commented):
We're in the process of merging feature-emac now, so we'll need fixes as soon as possible if this driver is going to make it to 5.9.0.

mikaleppanen commented on targets/TARGET_Realtek/TARGET_AMEBA/RTWInterface.cpp:
```diff
@@ -191,85 +179,106 @@ nsapi_error_t RTWInterface::connect()
+
+#define LWIP 0x11991199
+#if MBED_CONF_NSAPI_DEFAULT_STACK == LWIP
+    rtw_emac.wlan_emac_link_change(true);
```
rtw_emac.wlan_emac_link_change() calls can be removed from here and link status call can be made on emac power up.

M-ichae-l replied:
@mikaleppanen 
the `add_ethernet_interface` will run `emac_if_init` to set the callback and then `power up`. Logic here is correct. 
However, the `netif->flags` will be set by the following code. It is still cause the "connection error" from ` bringup()`.

mikaleppanen replied:
The netif->flags setting before power up was a bug in the lwip. I made a pull request to fix that: https://github.com/ARMmbed/mbed-os/pull/6926. 

You can remove the callback from there and take my pull request into your test build. It should work now correctly and trigger the link up from power up.

M-ichae-l replied:
Noted. Thank you.

mikaleppanen replied:
Actually it seems that for second connect after disconnect, the callback is needed. It could be unconditional before the interface check e.g. like this:

    rtw_emac.wlan_emac_link_change(true);

    if (!rtw_interface) {
        nsapi_error_t err = rtw_obn_stack.add_ethernet_interface(rtw_emac, true, &rtw_interface);
        if (err != NSAPI_ERROR_OK) {
            rtw_interface = NULL;
            return err;
        }
    }

M-ichae-l replied:
Thanks. Will add `rtw_emac.wlan_emac_link_change(true); `after` wifi_connect` .

mikaleppanen commented on targets/TARGET_Realtek/TARGET_AMEBA/RTWInterface.cpp:
```diff
@@ -191,85 +179,106 @@ nsapi_error_t RTWInterface::connect()
+#define TEST 0x33254234
+#if MBED_CONF_NSAPI_DEFAULT_STACK == TEST
+    if (rtw_if_bringup == NSAPI_ERROR_OK) {
+        rtw_emac.wlan_emac_link_change(true);
```
Also this can be removed.

mikaleppanen commented on targets/TARGET_Realtek/TARGET_AMEBA/rtw_emac.cpp:
```diff
@@ -73,80 +78,93 @@ static void wlan_get_hwaddr(emac_interface_t *emac, uint8_t *addr)
+bool RTW_EMAC::power_up()
 {
     wifi_on(RTW_MODE_STA);
     wait_ms(1000);
```
Link up can be made here. e.g. 
    if (emac_link_state_cb) {
        emac_link_state_cb(true);
    }

mikaleppanen replied:
emac_link_state_cb is the callback set by the stack, so it can be called here.

M-ichae-l replied:
Confirmed to use `wlan_emac_link_change(true); `
`wlan_emac_link_change(true);` and 
```
if (emac_link_state_cb) {
    emac_link_state_cb(true);
} 
```
actually are same. Thank you.

mikaleppanen commented on targets/TARGET_Realtek/TARGET_AMEBA/rtw_emac.h:
```diff
@@ -13,13 +13,155 @@
+    void *emac_link_state_data;
+    emac_link_input_cb_t emac_link_input_cb; 
+    emac_link_state_change_cb_t emac_link_state_cb; 
+    LWIPMemoryManager *memory_manager; 
```
This should be EMACMemoryManager

M-ichae-l replied:
Thank you. Confirmed, should use `EMACMemoryManager`.

mikaleppanen reviewed (commented):
Added some review comments

M-ichae-l committed:
rebase for feacture_emac

fix the conflicts with #6851.

M-ichae-l committed:
Merge branch 'feature-emac' into Add-rtl8195am-feature-emac

M-ichae-l committed:
updats add rtl8915am emac

Expecting the #6926
Corrected lwip netif flags to be set before power up

kjbracey commented on tools/test_configs/RealtekInterface.json:
```diff
@@ -9,7 +9,7 @@
         "connect-statement" : {
             "help" : "Disabled until EMAC updated",
-            "value" : "((RTWInterface *)net)->connect(WIFI_SSID, WIFI_PASSWORD, WIFI_SECURITY, WIFI_CHANNEL)"
+            "value" : "(net->wifiInterface()->connect(WIFI_SSID, WIFI_PASSWORD, WIFI_SECURITY, WIFI_CHANNEL)"
```
Unless I'm mistaken, this has an excess '(' at the front.

Could you put back the original help from master too:

    "help": "Must use 'net' variable name, replace WIFI_SSID, WIFI_PASSWORD, WIFI_SECURITY, WIFI_CHANNEL with your WiFi settings",

M-ichae-l replied:
Ok.

kjbracey commented on targets/TARGET_Realtek/TARGET_AMEBA/rtw_emac.cpp:
```diff
@@ -32,39 +33,42 @@
 
-static void wlan_get_ifname(emac_interface_t *emac, char *name, uint8_t size)
+uint32_t RTW_EMAC::get_align_preference() const
+{
```
This needs to return something. `1` is fine.

M-ichae-l replied:
Thanks. Confirmed to add `return true;` for `RTW_EMAC::get_align_preference() `.

kjbracey commented on targets/TARGET_Realtek/TARGET_AMEBA/RTWInterface.cpp:
```diff
@@ -19,12 +19,15 @@
 #include "wifi_conf.h"
-#include "lwip_stack.h"
 
+#include "LWIPStack.h"
```
You shouldn't be including `LWIPStack.h`

M-ichae-l replied:
Thanks. Confirmed to delete `LWIPStack.h` .

kjbracey commented on targets/TARGET_Realtek/TARGET_AMEBA/RTWInterface.cpp:
```diff
@@ -191,85 +179,93 @@ nsapi_error_t RTWInterface::connect()
     if (ret != RTW_SUCCESS) {
         printf("failed: %d\r\n", ret);
         return NSAPI_ERROR_NO_CONNECTION;
     }
```
As Mika mentioned, you will still need a "link up" callback here after the "wifi_connect", to ensure the upcall occurs if this is the second connect, and the stack is already bound to you. That's the direct mirror of the "link down" callback from "disconnect".

The callback in "power_up" then handles the case of the first connect, when the stack wasn't attached to hear this link-up callback.

M-ichae-l replied:
Thanks. Will add `rtw_emac.wlan_emac_link_change(true);` after `wifi_connect `.

M-ichae-l committed:
update add rtl8915am emac

add link call back after wifi_connect
add return for get_align_preference
delete LWIPStack.h from RTWInterface.cpp
coding style fix

kjbracey commented on tools/test_configs/RealtekInterface.json:
```diff
@@ -8,8 +8,8 @@
-            "help" : "Disabled until EMAC updated",
-            "value" : null
+            "help" : "Must use 'net' variable name, replace WIFI_SSID, WIFI_PASSWORD, WIFI_SECURITY, WIFI_CHANNEL with your WiFi settings",
+            "value" : "(net->wifiInterface()->connect(WIFI_SSID, WIFI_PASSWORD, WIFI_SECURITY, WIFI_CHANNEL)"
```
This still has a stray opening `'(` at the front. I can edit this myself - going to need to rebase and tidy this commit.

M-ichae-l replied:
Thank you!

kjbracey reviewed (approved):
Approved, but I'll rebase to tidy up and remove the opening '('.

It's passing tests here.

cmonr commented:
/morph build

cmonr commented:
@M-ichae-l Looks like all three compilers failed with a missing header against the REALTEK_RTL8195AM

`[DEBUG] Output:   #include "Rtw_emac.h"`

M-ichae-l commented:
@cmonr  the actual files is all in lowercase. Need to change all the` #include "Rtw_emac.h"` and `#include "Rtw_emac.cpp"` become `#include "rtw_emac.h"` and `#include "rtw_emac.cpp"` .

M-ichae-l committed:
update add rtl8915am emac

Change all the#include "Rtw_emac.h" and #include "Rtw_emac.cpp" become #include "rtw_emac.h" and #include "rtw_emac.cpp"
coding style fix

cmonr reviewed (approved):
👍

cmonr commented:
/morph build

kjbracey commented:
I had been trying to accelerate CI by combining the PRs here #6936 - the case fix is already made on that PR. I've stuck the "do not merge" label on the ones I've combined, so they don't need to be dealt with.

Unfortunately the CI has mostly been down within my working hours, so this hasn't really been working out.
